File: Napakalaki_Java/NapakalakiGame/Napakalaki.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;

import java.util.*;
/**
 *
 * @author l
 */
public class Napakalaki {

    Player currentPlayer;
    ArrayList<Player> players;
    CardDealer dealer;
    Monster currentMonster;
    int currentPlayerIndex;
 
    private static final Napakalaki instance = new Napakalaki();
    private Napakalaki(){
        currentPlayer=new Player("");
        players=new ArrayList();
        dealer=CardDealer.getInstance();
        currentMonster=null;
    };
    
    
    public static Napakalaki getInstance(){
        return instance;
    }
    
    private void initPlayers(ArrayList<String> names){
        for(String aux : names){
            
            Player player= new Player(aux);
            players.add(player);
        }
        

    }
    
    private Player nextPlayer(){
        if( currentPlayer.getHiddenTreasures().isEmpty() && currentPlayer.getVisibleTreasures().isEmpty()){
            Random rnum = new Random();
            return  players.get(rnum.nextInt(players.size()));
        }
        else{ if(players.indexOf(currentPlayer)==players.size()-1)
                return players.get(0);
              else return players.get(players.indexOf(currentPlayer)+1);
            }
            
    }
    
    private boolean nextTurnAllowed(){
        if(currentPlayer.validState())
            return true;
        else return false;
    }
    
    public static Napakalaki getInstace(){
        return instance;
    }
    
    public CombatResult developCombat(){
    
        CombatResult combatResult=currentPlayer.combat(currentMonster);
        
        dealer.giveMonsterBack(currentMonster);
        
        return combatResult;
    }
    
    public void discardVisibleTreasures(ArrayList<Treasure> treasures){
        
        
        for(Treasure treasure : treasures){
            currentPlayer.discardVisibleTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }
    
    public void discardHiddenTreasures(ArrayList<Treasure> treasures){

        for(Treasure treasure : treasures){
            currentPlayer.discardHiddenTreasure(treasure);
            dealer.giveTreasureBack(treasure);
        }
    }    
    
    public void makeTreasuresVisible(ArrayList<Treasure> treasures){
    
        for(Treasure t: treasures){
            currentPlayer.makeTreasureVisible(t);
        }
    }
    
    public boolean buyLevels( ArrayList<Treasure> visible, ArrayList<Treasure> hidden){
    
        boolean canI=currentPlayer.buyLevels(visible, hidden);
        return canI;
    }
    
    public void initGame( ArrayList<String> players){
        this.initPlayers(players);
        dealer.initCards();
        this.nextTurn();
        
    }

    public Player getCurrentPlayer(){
        return currentPlayer;
        
    }
    
    public Monster getCurrentMonster(){
        return currentMonster;
    }
    
    public boolean nextTurn(){
      boolean stateOk=this.nextTurnAllowed();
      if(stateOk){
          dealer=CardDealer.getInstance();
          currentMonster=dealer.nextMonster();
          currentPlayer=this.nextPlayer();
          
          boolean dead=currentPlayer.isDead();
          
          if(dead)
          {
              currentPlayer.initTreasures();
          }
          
      }
                return stateOk;
    }
    
    public boolean endOfTheGame(CombatResult result){
        if(result==CombatResult.WinAndWinGame)
            return true;
        else return false;
    }









}

File: Napakalaki_Ruby/Napakalaki.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

require_relative 'Card_dealer.rb'
require_relative 'Bad_consequence.rb'
require_relative 'CombatResult.rb'
require_relative 'dice.rb'
require_relative 'monster.rb'
#require_relative 'napakalaki.rb'
require_relative 'player.rb'
require_relative 'prize.rb'
require_relative 'treasure.rb'
require_relative 'treasure_kind.rb'

module NapakalakiGame
require "singleton" 

class Napakalaki

  include Singleton
  
  attr_reader :currentPlayer, :currentMonster
  def initialize
    @currentPlayer=Player.new("")
    @players=Array.new
    @dealer=CardDealer.instance
    @currentMonster=nil
  end
  
 
  def getCurrentPlayer
    return @currentPlayer
  end
  
  def getCurrentMonster
    return @currentMonster
  end
  #EXAMEN
  def nextTurnExam
    
     extraTreasures=Array.new
     extraTreasures=@currentPlayer.getExtraTreasures
     @currentMonster=dealer.nextMonster
     @currentPlayer.receiveTreasures(extraTreasures)
     nextPlayer()
     
  end
  #FINEXAMEN
  def developCombat
    
    combatResult=@currentPlayer.combat(@currentMonster)
    @dealer.giveMonsterBack(@currentMonster)
    
    return combatResult
    
    
  end
  
  def discardVisibleTreasure( treasures )
    treasures.each{ |treasure| @currentPlayer.discardVisibleTreasure(treasure) 
                                @dealer.giveTreasureBack(treasure)
    }
  end
  
  def discardHiddenTreasure( treasures )
    treasures.each{ |treasure| @currentPlayer.discardHiddenTreasure(treasure) 
                                @dealer.giveTreasureBack(treasure)
    }    
  end
  
  def makeTreasuresVisible( treasures )
    
    puts treasures.size
    treasures.each{ |t| @currentPlayer.makeTreasureVisible(t)}
    
  end
  
  def buyLevels( visible, hidden)
    canI=@currentPlayer.buyLevels(visible, hidden)
    return canI
  end
  
  def initGame( names )
    initPlayers(names)
    @dealer.initCards
    nextTurn
  end
  
  
  def nextTurn
    
    stateOk=nextTurnAllowed
    stateOk=@currentPlayer.validState
    
    if stateOk
      @currentMonster=@dealer.nextMonster
      @currentPlayer=nextPlayer
      
      dead=@currentPlayer.isDead
      #si es su primer turno no tiene tesoros y su nivel es 1
      if dead 
        @currentPlayer.initTreasures
      end
    end
    
    return stateOk
    
  end
  
  def endOfGame( result )
    return result==CombatResult::WinAndWinGame
    
  end
  
  private
  
  def initPlayers (names)
    names.each{
      |x| new_player=Player.new(x)
      @players << new_player
    }
  end
  
  def nextTurnAllowed
    return @currentPlayer.validState()
  end
  
  def nextPlayer
    if @currentPlayer.getVisibleTreasures.empty? and @currentPlayer.getHiddenTreasures.empty?
      return @players[rand(@players.size)]
    
    elsif(@players.index(@currentPlayer)==@players.size-1) 
          return @players[0]
        else
          return @players[@players.index(@currentPlayer)+1]
        end
    
    end
    
    
  end
  


end
